### DISCUSSION TITLE
Relationship using a self-join

### DISCUSSION THREAD

**Discussion by soylent** [sqlalchemy/sqlalchemy#9686] [Usage Questions] [answered]
How can I declaratively define a relationship on a SQLAlchemy model that joins the right table in the following manner:
```sql
SELECT *
FROM left_table left
JOIN left_table inter ON left.inter_id = inter.id
JOIN right_table right ON right.id = inter.right_id;
```
The culprit here is that the left table and the junction table are the same.

The following code generates a query that is close, but SQLAlchemy generates the wrong primaryjoin condition `... on left_table_1.inter_id = left_table_1.id` instead of `... on left_table_1.inter_id = left_table.id`. Am I missing something, or is this a bug?. I've tried using an alias table, but that didn't work. I'm using SQLAlchemy 1.3.24.

```python
class Left(db.Model):
    __tablename__ = 'left_table'

    id = db.Column(db.Integer, primary_key=True)
    inter_id = db.Column(db.Integer, db.ForeignKey('left_table.id'))
    right_id = db.Column(db.Integer, db.ForeignKey('right_table.id'))
    ...
    right = db.relationship(Right, secondary="left_table", lazy="joined", uselist=False)


class Right(db.Model):
    __tablename__ = 'right_table'

    id = db.Column(db.Integer, primary_key=True)
    ...
```

**Comment by zzzeek** [comment, marked as answer]
you can't use "secondary" in direct form, you'd want to use a form such as that indicated in https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#composite-secondary-joins and use an alias() of the table as added to "secondary", and you would need to mark this as viewonly=True as the ORM doesn't know how to persist this.

not a bug "secondary" can't be the same table as the host table, you have to alias it

**Reply by soylent** [reply]
Thanks, Mike. Your reply was very helpful. I ended up with the following code:
```python
Left.right = db.relationship(
    db.aliased(Right, db.join(db.aliased(Left), Right)),
    foreign_keys=Left.inter_id,
    lazy="joined",
    uselist=False,
    viewonly=True,
)

```
